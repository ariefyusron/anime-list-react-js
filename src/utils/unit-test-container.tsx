import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { ReactNode } from "react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

import themes from "@themes";

const queryClient = new QueryClient();

const renderWithTheme = (ui: ReactNode) => {
  return render(
    <ThemeProvider theme={themes}>
      <QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>
    </ThemeProvider>
  );
};

export { renderWithTheme };
