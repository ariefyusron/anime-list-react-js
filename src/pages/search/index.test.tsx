import { screen, fireEvent, within } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach } from "vitest";

import { renderWithTheme } from "../../utils/unit-test-container";

import SearchPage from ".";

const data = {
  pages: [
    {
      data: [
        {
          mal_id: 1,
          images: {
            webp: {
              image_url: "link.url"
            }
          },
          title: "Title",
          genres: [
            {
              name: "Action"
            },
            {
              name: "Comedy"
            }
          ],
          duration: "10 min",
          episodes: 10,
          type: "TV"
        },
        {
          mal_id: 2,
          images: {
            webp: {
              image_url: "link2.url"
            }
          },
          title: "Title2",
          genres: [
            {
              name: "Horor"
            },
            {
              name: "Comedy"
            }
          ],
          duration: "11 min",
          episodes: 11,
          type: "Movie"
        }
      ]
    }
  ]
};

let mockSearchValue = {
  isFetching: false,
  data: null,
  valueSearch: ""
};

const mockNavigate = vi.fn();
vi.mock("react-router", () => ({
  useNavigate: () => mockNavigate
}));

describe("SearchPage Component", () => {
  beforeEach(() => {
    vi.mock("../../hooks/useAnime", () => ({
      useGetAnimeListSearch: vi.fn(() => mockSearchValue)
    }));

    vi.clearAllMocks();
  });

  it("render isLoading Snapshot", () => {
    mockSearchValue = {
      isFetching: true,
      data: null,
      valueSearch: "test"
    };

    const component = renderWithTheme(<SearchPage />);

    expect(component).toMatchSnapshot();
  });

  it("render Not Found Snapshot", () => {
    mockSearchValue = {
      isFetching: false,
      data: null,
      valueSearch: "test"
    };

    const component = renderWithTheme(<SearchPage />);

    expect(component).toMatchSnapshot();
  });

  it("render Data Snapshot", () => {
    mockSearchValue = {
      isFetching: false,
      data: data.pages[0] as unknown as null,
      valueSearch: "test"
    };

    const component = renderWithTheme(<SearchPage />);

    expect(component).toMatchSnapshot();
  });

  it("should render the SearchPage component isLoading", () => {
    mockSearchValue = {
      isFetching: true,
      data: null,
      valueSearch: "test"
    };

    renderWithTheme(<SearchPage />);

    expect(screen.getByTestId("component-loader")).toBeInTheDocument();
  });

  it("should render the SearchPage component Not Found", () => {
    mockSearchValue = {
      isFetching: false,
      data: null,
      valueSearch: "test"
    };

    renderWithTheme(<SearchPage />);

    expect(screen.getByText(`"test" tidak ditemukan`)).toBeInTheDocument();
  });

  it("should render the SearchPage component render list data", () => {
    mockSearchValue = {
      isFetching: false,
      data: data.pages[0] as unknown as null,
      valueSearch: "test"
    };
    renderWithTheme(<SearchPage />);

    //item first
    const cardImageElementFirst = screen.getByTestId("1");
    const childElementFirst = within(cardImageElementFirst);
    const imageElementFirst = childElementFirst.getByAltText("image-anime");
    expect(imageElementFirst).toBeInTheDocument();
    expect(imageElementFirst).toHaveAttribute("src", "link.url");

    expect(childElementFirst.getByText("Title")).toBeInTheDocument();

    fireEvent.mouseEnter(cardImageElementFirst);
    expect(childElementFirst.getByText("TV")).toBeInTheDocument();
    expect(childElementFirst.getByText("10 episodes")).toBeInTheDocument();
    expect(childElementFirst.getByText("Action - Comedy")).toBeInTheDocument();

    fireEvent.click(cardImageElementFirst);
    expect(mockNavigate).toHaveBeenCalledWith("/detail/1");
    mockNavigate.mockClear();

    //item secondary
    const cardImageElementSecond = screen.getByTestId("2");
    const childElementSecond = within(cardImageElementSecond);
    const imageElementSecond = childElementSecond.getByAltText("image-anime");
    expect(imageElementSecond).toBeInTheDocument();
    expect(imageElementSecond).toHaveAttribute("src", "link2.url");

    expect(childElementSecond.getByText("Title2")).toBeInTheDocument();

    fireEvent.mouseEnter(cardImageElementSecond);
    expect(childElementSecond.getByText("Movie")).toBeInTheDocument();
    expect(childElementSecond.getByText("11 min")).toBeInTheDocument();
    expect(childElementSecond.getByText("Horor - Comedy")).toBeInTheDocument();

    fireEvent.click(cardImageElementSecond);
    expect(mockNavigate).toHaveBeenCalledWith("/detail/2");
    mockNavigate.mockClear();
  });
});
