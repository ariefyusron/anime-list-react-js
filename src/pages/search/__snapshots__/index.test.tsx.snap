// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchPage Component > render Data Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-jBIHhB eWMKmJ"
      >
        <div
          class="sc-gjcoXW jaiddG"
          data-testid="1"
        >
          <img
            alt="image-anime"
            class="sc-dVBluf drBQNX"
            src="link.url"
          />
          <div
            class="sc-fLDLck bpHFuX"
          >
            <h2
              class="sc-dycHyt hFOIiA"
            >
              Title
            </h2>
          </div>
        </div>
        <div
          class="sc-gjcoXW jaiddG"
          data-testid="2"
        >
          <img
            alt="image-anime"
            class="sc-dVBluf drBQNX"
            src="link2.url"
          />
          <div
            class="sc-fLDLck bpHFuX"
          >
            <h2
              class="sc-dycHyt hFOIiA"
            >
              Title2
            </h2>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-jBIHhB eWMKmJ"
    >
      <div
        class="sc-gjcoXW jaiddG"
        data-testid="1"
      >
        <img
          alt="image-anime"
          class="sc-dVBluf drBQNX"
          src="link.url"
        />
        <div
          class="sc-fLDLck bpHFuX"
        >
          <h2
            class="sc-dycHyt hFOIiA"
          >
            Title
          </h2>
        </div>
      </div>
      <div
        class="sc-gjcoXW jaiddG"
        data-testid="2"
      >
        <img
          alt="image-anime"
          class="sc-dVBluf drBQNX"
          src="link2.url"
        />
        <div
          class="sc-fLDLck bpHFuX"
        >
          <h2
            class="sc-dycHyt hFOIiA"
          >
            Title2
          </h2>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SearchPage Component > render Not Found Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-cMqPwU fppnNT"
      >
        <div>
          "
          test
          " tidak ditemukan
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-cMqPwU fppnNT"
    >
      <div>
        "
        test
        " tidak ditemukan
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SearchPage Component > render isLoading Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="component-loader"
        style="display: flex; justify-content: center; margin-top: 100px;"
      >
        <div
          class="sc-FEMpB jtZqmY"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="component-loader"
      style="display: flex; justify-content: center; margin-top: 100px;"
    >
      <div
        class="sc-FEMpB jtZqmY"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
